### ISSUE TITLE
ssdb slow start up time

### ISSUE THREAD

**Issue by nikos-glikis** [ideawu/ssdb#1349]
Hello,

SSDB is great, it has solved lots of the problems I have. Its the best server LevelDb server implementation so far.

There is one issue that is bothering me though. When for whatever reason ssdb closes, the start up time is hours on a modern server with SSD.

For example starting a 1.1TB data instance of SSDB takes ~2hours and it usually needs a lot more memory in the beginning. This is problematic for production large database use cases.

From what I understand, once SSDB starts it "reads" all keys. Is there any way we can speedup initial load times ? Even if that means that we use more disk space ?

Thanks

**Comment by ideawu** [comment]
A better way is to separate your data into several SSDB instances, each stores 300GB or so.

**Comment by nikos-glikis** [comment]
I don't use SSDB as a simple key-value store. I rely heavily on scans, so using many SSDB instances, is not an option. I would need a lot of client code to get the same functionality. Also we would have a lot more management problems for all the instances. And also it does not change the fact that I would need a lot of memory on boot and a lot of cores to support all the instances. 

I have made an experiment using https://github.com/tidwall/redcon and  https://github.com/syndtr/goleveldb. There is a 5%-10% performance hit when creating the keys, but the database starts immediately where areas SSDB takes hours to start and I have no surprises with the memory with large databases. From what I read, that is the way leveldb is supposed to work. It has a lot less features than SSDB, but it covers what I need (set, get, del, scan) I will end up going with this if I find no solution. 

Just to make clear, can I somehow avoid the slow startup, even if that means that I will loose some functionality ? SSDB is very stable and exactly what I need and I hate to not use it in production only for this issue!

**Comment by geert-hendrickx-be** [comment]
Do you regularly run `compact` on your database?  Specifically prior to a (planned) restart?

**Comment by nikos-glikis** [comment]
Yes. That helps a lot, but it still takes hours. Sometimes you cannot plan these things (server failure etc).

**Comment by geert-hendrickx-be** [comment]
Our db is only 3,5 GB, but we run compact weekly from cron and it makes a significant difference in startup times (from minutes to almost instant).  Even for unplanned restarts, compact will have run 'recently enough'.

**Comment by ideawu** [comment]
@nikos-glikis do you run SSDB on a rotational disk or a Solid State Disk?

**Comment by nikos-glikis** [comment]
For the same data SSD will take close to 1 hour closer to 40 minutes after compaction. But a rotational disk will take hours. 

But my opinion is that event if that was some minutes, its problematic for production environments. Wouldn't you agree ?

**Comment by ideawu** [comment]
agree. maybe the authors of leveldb would help, would you please post an issue on leveldb's issue list? https://github.com/google/leveldb

**Comment by nikos-glikis** [comment]
I have made a test with similar amount of data, leveldb + redcon and ssdb as I mentioned above.  SSDB took 10 seconds to load, leveldb + redon were instantly usable. Although I will need to test more, and the leveldb implementation is not in C (golang). I am trying to run bigger tests an the moment, I will let you know if I have something.

**Comment by nikos-glikis** [comment]
So I have ran an experement.

I can verify that with the crude server I have setup above (https://github.com/tidwall/redcon and https://github.com/syndtr/goleveldb), a database of ~700GB is accessible after ~12 seconds and that is on a HDD NOT an SSD. That is without any compaction or special care - I just closed and started again the executable. I have similar experience with smaller database created with the c version of leveldb. So my conclusion is that this is probably not a Leveldb issue.

**Comment by valeriansaliou** [comment]
Hi! I am seeing a similar slow-start issue (where SSDB would use 100% CPU upon start, and progressively eat all of the RAM), after manually restarting SSDB on a previously healthy and moderately-loaded server, on a DB of ~20GB on disk.

Note that this SSDB instance stores a lot of transient data (meaning, w/ a TTL set), thus most keys stored disappear automatically as they get expired after a few days. Also, a compaction is ran weekly. SSDB is replicated to a slave.

This slow/forever start only affects the write master. If I restart the read-only slave, it starts up immediately no matter what.

To fix the issue on my end, when a slow start event occurs (which could take hours, I'm not even sure this would complete), I'd need to flush the master, re-configure it as a read-only slave of the cluster slave, then re-replicate it, re-elect it to master, and restart it again). Only thereafter starts would be fast. Give it some weeks/months, and restarting the same master would be slow again.

I'm investigating SSDB's source code to find the potential curlprit, and will report my findings here.

**Comment by valeriansaliou** [comment]
Okay so I've investigated, and here are my raw notes on the findings. As a summary, it looks like this only happens for SSDB servers with a binlog enabled, configured to be the write master, running version `1.9.9` with a recently re-synchronized database from its slave (therefore, the binlog was restarted from zero a few weeks ago on `1.9.9`).

# Boot sequence analysis:

Some quick static analysis of the code lead to the following SSDB boot sequence:

1. main -> app.cpp
  | 1 -> Application::main
  | 2 -> Application::init
  | 3 -> Application::run

2. ssdb-server.cpp in Application::run()
  | 1. -> option.load()          <not_heavy>
  | 2. -> SSDB::open (data)      <potentially_heavy>
  | 3. -> SSDB::open (meta)      <potentially_heavy>
  | 4. -> NetworkServer::init()  <<NOT_REACHED?>>
  | 5. -> new SSDBServer()       <potentially_heavy>

3. serv.cpp in SSDBServer::SSDBServer()
  | 1. -> new BackendDump()        <maybe_okay>
  | 2. -> new BackendSync()        <maybe_okay>
  | 3. -> new ExpirationHandler()  <likely_curlprit>
  | 4. -> this->get_kv_range(()    <likely_curlprit>

---

# When starting ssdb:

Adding some breakpoint logs, re-compiling SSDB, and starting it to test a blocked start for real:

- Syslog shows:

```
Jun  2 14:27:22 test-ssdb-master-boot systemd[1]: Started SSDB Database.
Jun  2 14:27:22 test-ssdb-master-boot ssdb[1352]: ssdb-server 1.9.8
Jun  2 14:27:22 test-ssdb-master-boot ssdb[1352]: Copyright (c) 2012-2015 ssdb.io
Jun  2 14:27:22 test-ssdb-master-boot ssdb[1352]: 2022-06-02 14:27:22.149 [INFO ] ssdb-server.cpp(46): ssdb-server 1.9.8
Jun  2 14:27:22 test-ssdb-master-boot ssdb[1352]: 2022-06-02 14:27:22.149 [INFO ] ssdb-server.cpp(47): conf_file        : /etc/ssdb.conf
Jun  2 14:27:22 test-ssdb-master-boot ssdb[1352]: 2022-06-02 14:27:22.149 [INFO ] ssdb-server.cpp(48): log_level        : debug
Jun  2 14:27:22 test-ssdb-master-boot ssdb[1352]: 2022-06-02 14:27:22.149 [INFO ] ssdb-server.cpp(49): log_output       : stdout
Jun  2 14:27:22 test-ssdb-master-boot ssdb[1352]: 2022-06-02 14:27:22.149 [INFO ] ssdb-server.cpp(50): log_rotate_size  : 0
Jun  2 14:27:22 test-ssdb-master-boot ssdb[1352]: 2022-06-02 14:27:22.149 [INFO ] ssdb-server.cpp(52): main_db          : /var/lib/ssdb/data
Jun  2 14:27:22 test-ssdb-master-boot ssdb[1352]: 2022-06-02 14:27:22.149 [INFO ] ssdb-server.cpp(53): meta_db          : /var/lib/ssdb/meta
Jun  2 14:27:22 test-ssdb-master-boot ssdb[1352]: 2022-06-02 14:27:22.149 [INFO ] ssdb-server.cpp(54): cache_size       : 1024 MB
Jun  2 14:27:22 test-ssdb-master-boot ssdb[1352]: 2022-06-02 14:27:22.149 [INFO ] ssdb-server.cpp(55): block_size       : 32 KB
Jun  2 14:27:22 test-ssdb-master-boot ssdb[1352]: 2022-06-02 14:27:22.149 [INFO ] ssdb-server.cpp(56): write_buffer     : 32 MB
Jun  2 14:27:22 test-ssdb-master-boot ssdb[1352]: 2022-06-02 14:27:22.149 [INFO ] ssdb-server.cpp(57): max_open_files   : 500
Jun  2 14:27:22 test-ssdb-master-boot ssdb[1352]: 2022-06-02 14:27:22.149 [INFO ] ssdb-server.cpp(58): compaction_speed : 100 MB/s
Jun  2 14:27:22 test-ssdb-master-boot ssdb[1352]: 2022-06-02 14:27:22.149 [INFO ] ssdb-server.cpp(59): compression      : yes
Jun  2 14:27:22 test-ssdb-master-boot ssdb[1352]: 2022-06-02 14:27:22.149 [INFO ] ssdb-server.cpp(60): binlog           : yes
Jun  2 14:27:22 test-ssdb-master-boot ssdb[1352]: 2022-06-02 14:27:22.149 [INFO ] ssdb-server.cpp(61): binlog_capacity  : 20000000
Jun  2 14:27:22 test-ssdb-master-boot ssdb[1352]: 2022-06-02 14:27:22.149 [INFO ] ssdb-server.cpp(62): sync_speed       : -1 MB/s
```

- And then it freezes on added breakpoint:

```
Jun  2 15:04:00 test-ssdb-master-boot ssdb[18430]: 2022-06-02 15:04:00.568 [INFO ] ssdb-server.cpp(67): will open ssdb -> data
Jun  2 15:04:00 test-ssdb-master-boot ssdb[18430]: 2022-06-02 15:04:00.568 [INFO ] ssdb_impl.cpp(39): (ssdb_impl) begin
Jun  2 15:04:00 test-ssdb-master-boot ssdb[18430]: 2022-06-02 15:04:00.568 [INFO ] ssdb_impl.cpp(43): (ssdb_impl) constructed
Jun  2 15:04:00 test-ssdb-master-boot ssdb[18430]: 2022-06-02 15:04:00.568 [INFO ] ssdb_impl.cpp(59): (ssdb_impl) configured
Jun  2 15:04:00 test-ssdb-master-boot ssdb[18430]: 2022-06-02 15:04:00.568 [INFO ] ssdb_impl.cpp(63): (ssdb_impl) pre-leveldb open
Jun  2 15:04:00 test-ssdb-master-boot ssdb[18430]: 2022-06-02 15:04:00.580 [INFO ] ssdb_impl.cpp(67): (ssdb_impl) post-leveldb open
Jun  2 15:04:00 test-ssdb-master-boot ssdb[18430]: 2022-06-02 15:04:00.580 [INFO ] ssdb_impl.cpp(74): (ssdb_impl) pre-binlog construct
Jun  2 15:04:00 test-ssdb-master-boot ssdb[18430]: 2022-06-02 15:04:00.580 [INFO ] binlog.cpp(160): (binlog) construct
Jun  2 15:04:00 test-ssdb-master-boot ssdb[18430]: 2022-06-02 15:04:00.580 [INFO ] binlog.cpp(169): (binlog) configured
Jun  2 15:04:00 test-ssdb-master-boot ssdb[18430]: 2022-06-02 15:04:00.580 [INFO ] binlog.cpp(175): (binlog) start scan
Jun  2 15:04:00 test-ssdb-master-boot ssdb[18430]: 2022-06-02 15:04:00.580 [INFO ] binlog.cpp(303): (binlog) find_min -> begin
Jun  2 15:04:00 test-ssdb-master-boot ssdb[18430]: 2022-06-02 15:04:00.580 [INFO ] binlog.cpp(308): (binlog) find_min -> pre-iterator
Jun  2 15:04:00 test-ssdb-master-boot ssdb[18430]: 2022-06-02 15:04:00.580 [INFO ] binlog.cpp(310): (binlog) find_min -> pre-seek
```

- Which corresponds to changed code ssdb-server.cpp:67 with:

```cpp
log_info("will open ssdb -> data"); // <-- added log breakpoint
data_db = SSDB::open(option, data_db_dir);
```

- While the `pre-binlog construct` line corresponds to ssdb_impl.cpp:76, right before:

```cpp
ssdb->binlogs = new BinlogQueue(ssdb->ldb, opt.binlog, opt.binlog_capacity);
```

- Which itself blocks right before binlog.cpp:178, right at:

```
Binlog log;
if(this->find_last(&log) == 1){
  this->last_seq = log.seq();
}
```

- `find_last` calls `find_min` internally, which has the following code w/ breakpoints:

```cpp
int BinlogQueue::find_min(Binlog *log) const{
  log_info("(binlog) find_min -> begin");

  int ret = 0;
  std::string key_str = encode_seq_key(0);
  leveldb::ReadOptions iterate_options;
  log_info("(binlog) find_min -> pre-iterator");
  leveldb::Iterator *it = db->NewIterator(iterate_options);
  log_info("(binlog) find_min -> pre-seek");
  it->Seek(key_str); // <-- THIS IS THE LINE BLOCKING
  log_info("(binlog) find_min -> pre-valid");
  if(it->Valid()){
    leveldb::Slice key = it->key();
    if(decode_seq_key(key) != 0){
      leveldb::Slice val = it->value();
      log_info("(binlog) find_min -> pre-load");
      if(log->load(val) == -1){
        ret = -1;
      }else{
        ret = 1;
      }
    }
  }
  log_info("(binlog) find_min -> pre-destruct");
  delete it;
  return ret;
}
```

- Then CPU spikes to 100%
- netstat -tupln shows no ssdb listener yet (cause its not yet called the method to bind to the network)
- iotop -o shows that the disk read activity for ssdb is going at 100MB/s, while no disk write shows (compaction size was set to 5MB/s max, so its not that)

---

# Findings:

- The ChangeLog file shows version `1.9.7` had a bugfix of `Fix startup blocked by BinlogQueue::find_last()(2018-08-08)`, which I suppose is related to this issue. However, even when running 1.9.8, this still occurs. Commit: https://github.com/ideawu/ssdb/commit/785221cc68dac0a0fe0547f6b6fdf008bbbadf54
- 1.9.9 similarly shows: `Fix binlog resource leak bug(2020-09-25)`. Commit: https://github.com/ideawu/ssdb/commit/c0c3bca0aca5a6a8b5c2a98ba1b966df4e719650
- binlog.cpp:182 has a line commented regarding: `this->clean_obsolete_binlogs();`, looks like we could need this to keep binlogs healthy? Its never called across all the code, so pbb binlogs get HUGE and therefore the boot scan takes forever?

**Comment by valeriansaliou** [comment]
Could this help? https://github.com/google/leveldb/releases/tag/1.23 ; see:

> Optimize leveldb block seeks to utilize the current iterator location. This is beneficial when iterators are reused and seeks are not random but increasing. It is additionally beneficial with larger block sizes and keys with common prefixes.

SSDB is currently using LevelDB `1.20`.
